### 1. Create a BED file with genomic position of Pseudogene and CDS regions.


### 2. Annotation of our samples with BED.file 


### 3. Merge VCF (Chr 22) of multiple samples into one VCF 


### 4. Evaluate the freq of allele with CSQ in VCF_merged in our population


### 5. Use a tool to convert a VCF into a TSV.file for import in R
